### PR TITLE
`useRecordContext` may return `undefined`

### PR DESCRIPTION
**Pull request by KoltesDigital**
[The doc mentions it](https://marmelab.com/react-admin/doc/4.0/useRecordContext.html#optimistic-rendering) but TS doesn't enforce it. I believe it should.

### CODE REVIEW

**Review by djhi — APPROVED** [review]
Thanks!

**Review by slax57 — APPROVED** [review]
Thank you for this submission, this is a good idea indeed!
I'll just wait for the rest of the team to give their approval before merging 🙂

**Review by WiXSL — APPROVED** [review]

**Review by fzaninotto — APPROVED** [review]

**Comment by fzaninotto** [comment]
Thanks!

**Comment by KoltesDigital** [comment]
That was quick, thanks :)

**Pull request merged**
